### PR TITLE
New module for live backgrounds

### PR DESCRIPTION
**Pull request by phlash**
New code module to handle backgrounds, and in particular live backgrounds from local files or network streams.

Partly addresses #70, in that backgrounds are now a more flexible item, although no UI is present to select one while running.
Also partly addresses #43, in that video backgrounds are now possible.

### CODE REVIEW

File: app/background.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    std::thread thr;
+    std::mutex mux;
+};
+
+// Internal video reader thread
+static void read_thread(background_t *pbkd) {
+    if (pbkd->dbg) fprintf(stderr, "background: thread start\n");
+    struct timespec last;
+    struct timespec proc;
+    clock_gettime(CLOCK_REALTIME, &last);
```
Comment (BenBE) [inline app/background.cc:31]: std::chrono -> monotonic_clock

File: app/background.cc [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    proc = last;
+    while (pbkd->run) {
+        // read frame under lock
+        pbkd->mux.lock();
```
Comment (BenBE) [inline app/background.cc:35]: Never unlocked if read feils

Comment (BenBE) [reply app/background.cc:35]: Also use RAII locks via std::lock_guard here if possible.

File: app/background.cc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                cv::resize(pbkd->raw, frame, cv::Size(240,160));
+                cv::putText(frame, msg, cv::Point(5,frame.rows-5), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(255,255,255));
+                cv::imshow("Background", frame);
+            }
+            pbkd->mux.unlock();
+            // wait for next frame, some sources are real-time, others are not, this ensures all play in real-time.
+            if (pbkd->frm>1) {
+                long adj = (now.tv_sec-proc.tv_sec)*1000000000l + (now.tv_nsec-proc.tv_nsec);
+                now.tv_sec = 0;
+                now.tv_nsec= (long)(1.0/pbkd->fps*1e9);
```
Comment (BenBE) [inline app/background.cc:57]:
```suggestion
                now.tv_nsec= (long)(1.0e9/pbkd->fps);
```

File: app/background.cc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+void *load_background(const char *path, int debug) {
+    background_t *pbkd = new background_t;
+    background_t &bkd = *pbkd;
+
+    bkd.dbg = debug;
+    bkd.cap.open(path, cv::CAP_ANY);    // explicitly ask for auto-detection of backend
+    if (!bkd.cap.isOpened()) {
+        if (bkd.dbg) fprintf(stderr, "cv::VideoCapture cannot open: %s\n", path);
+        delete pbkd;
```
Comment (BenBE) [inline app/background.cc:90]: Use try/catch to ensure memory is freed on error

File: app/background.cc [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    // Here be the logic...
+    //  if: can read 2 video frames => it's a video
+    //  else: is loaded as an image => it's an image
+    //  else: it's not usable.
+    if (bkd.cap.read(bkd.raw) && bkd.cap.read(bkd.raw)) {
+        // it's a video, try a reset and start reader thread..
+        if (bkd.cap.set(cv::CAP_PROP_POS_FRAMES, 0))
+            bkd.frm = 0;
+        else
+            bkd.frm = 2;
```
Comment (BenBE) [inline app/background.cc:106]:
```suggestion
        bkd.frm = bkd.cap.set(cv::CAP_PROP_POS_FRAMES, 0) ? 0 : 2;
```

Apart from this explain those magic numbers …

File: app/background.cc [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    } else {
+        // resize still image as requested into out
+        cv::resize(pbkd->raw, out, cv::Size(width, height));
+        frm = 1;
+    }
+    return frm;
+}
+
+void drop_background(void *handle) {
+    if (handle) {
```
Comment (BenBE) [inline app/background.cc:146]: Negate condition to exit early. Reduces indentation for the proc overall.

File: app/background.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/* This is licensed software, @see LICENSE file.
+ * Authors - @see AUTHORS file. */
+
+#ifndef _BACKGROUND_H_
+#define _BACKGROUND_H_
+
+#include <opencv2/core/mat.hpp>
+
```
Comment (BenBE) [inline app/background.h:8]: Forward-declare `struct background_t;` to pass typed pointers to all those functions. Doesn't affect actual API/structure for this structure on actual definition.

File: app/deepseg.cc [hunk 8]
```diff
@@ -454,18 +458,24 @@ int main(int argc, char* argv[]) try {
...
 			ai.get_output_mask(mask);
 
+			// get background frame
+			if (pbk) {
+				grab_background(pbk, width, height, bg);
+			} else {
+				if (blur_strength) {
+					raw.copyTo(bg);
+					cv::GaussianBlur(bg,bg,cv::Size(blur_strength,blur_strength),0);
+				}
```
Comment (BenBE) [inline app/deepseg.cc:472]: Any reason blurring the background is skipped for animated backgrounds? IMHO there's no real reason why this shouldn't be possible.

File: app/deepseg.cc [hunk 9]
```diff
@@ -403,17 +409,15 @@ int main(int argc, char* argv[]) try {
 	printf("back:   %s\n", back ? back : "(none)");
 	printf("model:  %s\n\n", modelname);
 
-	cv::Mat bg;
-	if (back) {
-		bg = cv::imread(back);
-	}
-	if (bg.empty()) {
+	void *pbk = (back) ? load_background(back, debug) : nullptr;
```
Comment (BenBE) [inline app/deepseg.cc:412]:
```suggestion
	std::shared_ptr<background_t> pbk(back ? load_background(back, debug) : nullptr, [](background_t* p) {
		if(p)
			drop_background(p);
	});
```

**Review by BenBE — CHANGES_REQUESTED** [review]

File: app/background.cc (hunk 1 above)
Comment (phlash) [reply app/background.cc:31]: Yeah - I cut/past that from test code :smile: will fix.

File: app/background.cc (hunk 3 above)
Comment (phlash) [reply app/background.cc:63]: Heh, yep, will fix

File: app/background.cc (hunk 4 above)
Comment (phlash) [reply app/background.cc:94]: Good point - yep, will fix.

File: app/background.cc (hunk 5 above)
Comment (phlash) [reply app/background.cc:110]: I will add a suitable comment!

File: app/background.cc (hunk 6 above)
Comment (phlash) [reply app/background.cc:150]: Thanks - yep, will fix.

File: app/background.h (hunk 7 above)
Comment (phlash) [reply app/background.h:8]: Ah yep, will fix.

File: app/deepseg.cc (hunk 9 above)
Comment (phlash) [reply app/deepseg.cc:412]: Yep, along with typing as earlier.

File: app/deepseg.cc (hunk 8 above)
Comment (phlash) [reply app/deepseg.cc:472]: Good point, I think the original logic was that you have a background exactly as you would like it (still or animated), or you wish to blur the original video background and use that. But possibly a poor assumption, we can blur anything :smile:

File: app/background.cc (hunk 2 above)
Comment (phlash) [reply app/background.cc]: Yep, spotted that - fell in my own trap! Now using RAII locks.

File: app/background.cc (hunk 1 above)
Comment (BenBE) [reply app/background.cc:31]: Especially note the difference between wall clock time and monotonic clock time …

Comment (phlash) [reply app/background.cc:31]: decided to use `steady_clock` (monotonic)

File: app/background.cc [hunk 10]
```diff
@@ -24,45 +23,45 @@ struct background_t {
...
-                    nanosleep(&now, nullptr);
-                } else {
-                    if (pbkd->dbg) fprintf(stderr, "background: wait=%ld adj=%ld\n", now.tv_nsec, adj);
+                auto adj = std::chrono::duration_cast<std::chrono::nanoseconds>(now-proc);
+                auto del = std::chrono::nanoseconds((long)(1e9/pbkd->fps));
+                // adjust for processing time and sleep or skip if no time left
+                if (del > adj) {
+                    del -= adj;
+                    std::this_thread::sleep_for(del);
                 }
```
Comment (BenBE) [inline app/background.cc]: Prone to overall jitter AFAIC. Especially as the `std::this_thread::sleep_for` does not guarantee exact timings.

File: app/background.cc [hunk 11]
```diff
@@ -79,52 +78,56 @@ static void read_thread(background_t *pbkd) {
...
-    } else {
-        // static image file, try loading..
-        bkd.cap.release();
-        bkd.raw = cv::imread(path);
-        bkd.vid = false;
-        if (bkd.raw.empty()) {
-            if (bkd.dbg) fprintf(stderr, "cv::imread cannot read: %s\n", path);
-            delete pbkd;
+std::shared_ptr<background_t> load_background(const char *path, int debug) {
+    std::shared_ptr<background_t> pbkd(new background_t);
```
Comment (BenBE) [inline app/background.cc:82]:
```suggestion
    std::shared_ptr<background_t> pbkd = std::make_shared<background_t>();
```

Even might:
```suggestion
    auto pbkd = std::make_shared<background_t>();
```

As the return type of `std::make_shared<T>` is `std::shared_ptr<T>` …

Also note the proposed dtor trick I commented yesterday.

File: app/background.cc [hunk 12]
```diff
@@ -79,52 +78,56 @@ static void read_thread(background_t *pbkd) {
...
+            pbkd->cap.release();
+            pbkd->raw = cv::imread(path);
+            pbkd->vid = false;
+            if (pbkd->raw.empty()) {
+                if (pbkd->dbg) fprintf(stderr, "background: imread cannot open: %s\n", path);
+                return nullptr;
+            }
+        }
+        if (pbkd->dbg) fprintf(stderr, "background properties:\n\tvid: %s\n\tfcc: %08x (%.4s)\n\tfps: %f\n\tcnt: %d\n",
+            pbkd->vid ? "yes":"no", fcc, (char *)&fcc, pbkd->fps, cnt);
```
Comment (BenBE) [inline app/background.cc:118]:
```suggestion
        if (pbkd->dbg)
                fprintf(stderr, "background properties:\n\tvid: %s\n\tfcc: %08x (%.4s)\n\tfps: %f\n\tcnt: %d\n",
                        pbkd->vid ? "yes" : "no", fcc, (char *)&fcc, pbkd->fps, cnt);
```

**Comment by floe** [comment]
General question: would it maybe make sense to use GStreamer instead for the whole video streaming thing? (IIRC you can use GStreamer pipelines as a video capture device in OpenCV, although I'm not 100% sure anymore about how.)

**Comment by BenBE** [comment]
Would be interesting to take the contents of another window as background … Not sure if GStreamer allows for that (but IIRC it somehow did) …

File: app/background.cc (hunk 10 above)
Comment (phlash) [reply app/background.cc]: Any thoughts on what to use from `std::` to improve the timing? I don't think there are any underlying implementations that provide guarantees on resolution, hence nothing in the std library can either?

File: app/background.cc (hunk 11 above)
Comment (phlash) [reply app/background.cc:82]: Ah - I missed the dtor trick, that makes sense now I've re-read it!

**Comment by phlash** [comment]
> General question: would it maybe make sense to use GStreamer instead for the whole video streaming thing? (IIRC you can use GStreamer pipelines as a video capture device in OpenCV, although I'm not 100% sure anymore about how.)

I've looked at GStreamer a little, while considering a module to connect libbackscrub into the ecosystem, it's quite invasive to the programming style (ie: you end up writing a GApplication and supporting paraphernalia), so I wouldn't be keen to adapt this codebase (it would be a re-write), instead continue with creating a module that other GStreamer apps can use.

File: app/background.cc (hunk 10 above)
Comment (phlash) [reply app/background.cc]: I've introduced some min/max measures, and switched to absolute target time via `std::this_thread::sleep_util()` calculated from frame rate. This doesn't show visible jitter (to me anyway!) with a smoothly moving video (eg: https://file-examples-com.github.io/uploads/2020/03/file_example_WEBM_1920_3_7MB.webm). would still welcome thoughts :smile:

File: app/background.cc [hunk 13]
```diff
@@ -45,23 +46,24 @@ static void read_thread(std::shared_ptr<background_t> pbkd) {
...
-            if (pbkd->frm>1) {
-                auto adj = std::chrono::duration_cast<std::chrono::nanoseconds>(now-proc);
-                auto del = std::chrono::nanoseconds((long)(1e9/pbkd->fps));
-                // adjust for processing time and sleep or skip if no time left
-                if (del > adj) {
-                    del -= adj;
-                    std::this_thread::sleep_for(del);
-                }
-                proc = std::chrono::steady_clock::now();
+            proc += std::chrono::nanoseconds((long)(1e9/pbkd->fps));
```
Comment (BenBE) [inline app/background.cc:63]: What about dropping frames when too slow?

File: app/background.cc [hunk 14]
```diff
@@ -78,10 +80,32 @@ static void read_thread(std::shared_ptr<background_t> pbkd) {
...
+        // clean up
+        pbkd->raw.release();
+    }
+    delete pbkd;
+}
+
 std::shared_ptr<background_t> load_background(const char *path, int debug) {
-    std::shared_ptr<background_t> pbkd(new background_t);
+    // allocate a shared pointer around storage for the handle, associate custom deleter to clean up when eventually released
+    auto pbkd = std::shared_ptr<background_t>(new background_t, [](background_t *p) { drop_background(p); });
```
Comment (BenBE) [inline app/background.cc:104]:
```suggestion
    auto pbkd = std::shared_ptr<background_t>(new background_t, drop_background });
```

File: app/background.cc (hunk 13 above)
Comment (phlash) [reply app/background.cc:63]: Good question - if we run out of cycles simply reading frames.. we have bigger problems? Difficult to drop frames on network streams (especially if they are http-like streams), so non-trivial to do this I think?

File: app/background.cc (hunk 14 above)
Comment (phlash) [reply app/background.cc:104]: heh, of course :smile:

**Review by BenBE — APPROVED** [review]

**Comment by phlash** [comment]
..and now I remember that the usage text should probably change :smile: bug fix incoming!

**Pull request merged**